core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala

```diff
@@ -460,25 +492,195 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+                 '<br>Task Deserialization Time: ${UIUtils.formatDuration(deserializationTime)}' +
+                 '<br>Shuffle Read Time: ${UIUtils.formatDuration(shuffleReadTime)}' +
+                 '<br>Executor Computing Time: ${UIUtils.formatDuration(executorComputingTime)}' +
+                 '<br>Shuffle Write Time: ${UIUtils.formatDuration(shuffleWriteTime)}' +
+                 '<br>Result Serialization Time: ${UIUtils.formatDuration(serializationTime)}' +
+                 '<br>Getting Result Time: ${UIUtils.formatDuration(gettingResultTime)}">' +
+                 '<svg class="task-assignment-timeline-duration-bar">' +
+                 '<rect class="scheduler-delay-proportion" ' +
+                   'x="${schedulerDelayProportionPos}%" y="0px" height="26px"' +
+                   'width="${schedulerDelayProportion}%""></rect>' +
+                 '<rect class="deserialization-time-proportion" '+
+                   'x="${deserializationTimeProportionPos}%" y="0px" height="26px"' +
+                   'width="${deserializationTimeProportion}%"></rect>' +
+                 '<rect class="shuffle-read-time-proportion" ' +
+                   'x="${shuffleReadTimeProportionPos}%" y="0px" height="26px"' +
+                   'width="${shuffleReadTimeProportion}%"></rect>' +
+                 '<rect class="executor-runtime-proportion" ' +
+                   'x="${executorRuntimeProportionPos}%" y="0px" height="26px"' +
+                   'width="${executorComputingTimeProportion}%"></rect>' +
+                 '<rect class="shuffle-write-time-proportion" ' +
+                   'x="${shuffleWriteTimeProportionPos}%" y="0px" height="26px"' +
+                   'width="${shuffleWriteTimeProportion}%"></rect>' +
+                 '<rect class="serialization-time-proportion" ' +
+                   'x="${serializationTimeProportionPos}%" y="0px" height="26px"' +
+                   'width="${serializationTimeProportion}%"></rect>' +
+                 '<rect class="getting-result-time-proportion" ' +
+                   'x="${gettingResultTimeProportionPos}%" y="0px" height="26px"' +
+                   'width="${gettingResultTimeProportion}%"></rect></svg>',
+               'start': new Date(${launchTime}),
+               'end': new Date(${finishTime})
+             }
+           """
+        timelineObject
+      }.mkString("[", ",", "]")
+
+    val groupArrayStr = executorsSet.map {
+      case (executorId, host) =>
+        s"""
+            {
+              'id': '${executorId}',
+              'content': '${executorId} / ${host}',
+            }
+          """
+    }.mkString("[", ",", "]")
+
+    val maxZoom = maxFinishTime - minLaunchTime
+    <span class="expand-task-assignment-timeline">
+      <span class="expand-task-assignment-timeline-arrow arrow-closed"></span>
+      <a>Event Timeline {
+        if (MAX_TIMELINE_TASKS <= numEffectiveTasks) {
+          s"(Most recent ${MAX_TIMELINE_TASKS})"
```

Comment:
As a usability thing, I actually think it would be better to always have this just say "Event Time", and then when there are more than MAX_TIMELINE_TASKS tasks, as the first thing when you expand the timeline, have a bold-ed warning that says something like "This stage has more than the maximum number of tasks that can be shown in the visualization! Only the first 1000 tasks (of $numEffectiveTasks total) are shown."